### ISSUE TITLE
[PROPOSAL] Adapt precision to mesh bounds

### ISSUE THREAD

**Issue by gentlecolts** [CGCookie/retopoflow#1254]
**PLEASE READ THE FOLLOWING AND USE THE TEMPLATE TO POST YOUR ISSUE**

*Questions not directly related to the RetopoFlow project are considered general and should be posted elsewhere (e.g. git, blender, python, etc).*

**Before raising the issue, please check the following first:**

- [x] Read through the README.md file in the branch used to assure the process taken is correct
- [x] Check the existing [issues](https://github.com/CGCookie/retopoflow/issues) and [pull requests](https://github.com/CGCookie/retopoflow/pulls) to make sure the issue has not already been reported and/or fixed.
- [x] You can replicate the bug and will provide as much info as possible with blend files, screenshots, logfiles, as well as show the expected result and actual result.

&nbsp;

Retopoflow Version: 3.4.0

Blender Version/Hash: 3.6.0

Platform Version/Distribution: all

&nbsp;

**Issue:**

Retopoflow currently exhibits poor user experience when operating on physically small elements.  This includes poor selection of default clipping and "factor" values, poor selection of mirror threshold, visual artifacting on symmetry "edge" display mode, etc.  Retoppoflow should consider the absolute scale when for these elements, and adjust its own thresholds accordingly

&nbsp;

**How to Reproduce:**

- operate on small element, perhaps a pendant or other accessory, observe issues described

**Comment by jlampel** [comment]
RetopoFlow actually resizes all meshes to fit in the same bounds to avoid exactly this issue and a few others. Blender has changed a lot since this scale workaround was necessary, so it may be something we can skip in the future. In the meantime, my guess is that the unit scale of your scene (Scene Properties -> Units -> Unit Scale) is set to something very small or very large. Can you try setting that to 1 and seeing if that helps?

**Comment by gentlecolts** [comment]
In my case, I had not changed the default scene units.  I can very easily reproduce this less than desirable behavior as well.  In the attached blend file, give retopoflow a try with the attached scene, created by applying a large amount of subdiv to a cube, shrinking it, applying scale, and moving it away from origin.

[retopoflow 3.4.0 scale issue.zip](https://github.com/CGCookie/retopoflow/files/12364294/retopoflow.3.4.0.scale.issue.zip)

**Comment by vxlcoder** [comment]
ah....  in the example you provided, @gentlecolts, the tiny, off-origin mesh has precision problems for sure. i think that accounts for most of the issues you noted.

When RF starts, we sneakily scales the objects behind the scenes to a more "normalized" size (based on bounding box) so that RF does not need to think about scale as much (many problems become much easier to solve when everything is a normal size).  however, we do _not_ offset the objects so the midpoint of bbox is at origin, and this is what is causing problems.  in your example, the object is super small, so RF will scale everything by ~596.  since the origin of the cube is originally at (0,0,1.9618)m, after scaling the origin is now at (0,0,1168.4)m, and float/double maths lose precision when far away from 0.

when the object is moved to (0,0,0), then RF works as expected even if the scale is not modified.

In your original .blend file (not the one you posted here), does RF work as expected if you move the source object(s) to the origin before starting?

a couple of notes to myself:

- i did notice a few visual bugs with the world unit scale is non-unit with this small object... i think it's just the clipping planes are not getting set correctly.  but also if the source mesh is scaled, starting RF At Active will copy that scaling, and the scaling affects the normal offset of the target visualization, so all the verts could be far from the surface)
- i have a few other thoughts on how RF can normalize the scene.  i wonder if one of them might be a better option.

**Comment by gentlecolts** [comment]
I seem to observe seemingly equally incorrect behavior (bad selection of clip planes, unusable "edge" view for symmetry, etc) regardless of mesh location, scale, orientation, etc, both in my original scene and the test case I provided.  I will add that I am on linux (manjaro) with nvidia drivers, so it's entirely possible that my observation of the symptoms of incorrect behavior here may differ than on other platforms

**Comment by vxlcoder** [comment]
thanks for pointing that out (again), @gentlecolts.  i do notice that now.

i focused on the precision errors of floating point numbers when far from 0 and missed your notes on the symmetry and clip.  will dig into that.

**Comment by jlampel** [comment]
This is handled very differently now in v4. The precision for merging and mirroring is handled natively and seems to work as expected. The tools themselves seem to behave fine. 

Working at extreme scales has always been awkward in Blender and takes some setup, and RF is no exception. Perhaps there is more we could do to make this more automatic in the future. But the more we do that, the more we deviate from Blender's native settings, which can also cause confusion. 

I'll close this out as the core issue is no longer relevant, but please feel free to open up a new issue if you find any particular tool misbehaving at extreme scales in v4. Unfortunately it is just too big of a task to add it to v3 at this point.
